File: windows/SqueezeBar/Models/StagedQueueItem.cs
using System;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;

namespace SqueezeBar.Models;

public class StagedQueueItem : INotifyPropertyChanged
{
    public Guid Id { get; } = Guid.NewGuid();
    public string FilePath { get; }
    public long OriginalSize { get; }
    public MediaType MediaType { get; }

    private double _customQuality = 0.80;
    public double CustomQuality
    {
        get => _customQuality;
        set { _customQuality = value; OnPropertyChanged(); OnPropertyChanged(nameof(CustomSettingSummary)); }
    }

    private double _customResolutionScale = 1.0;
    public double CustomResolutionScale
    {
        get => _customResolutionScale;
        set { _customResolutionScale = value; OnPropertyChanged(); OnPropertyChanged(nameof(CustomSettingSummary)); }
    }

    private ImageFormatPolicy _customImageFormat = ImageFormatPolicy.PreserveOriginal;
    public ImageFormatPolicy CustomImageFormat
    {
        get => _customImageFormat;
        set { _customImageFormat = value; OnPropertyChanged(); OnPropertyChanged(nameof(CustomSettingSummary)); }
    }

    private VideoCodecPreference _customVideoCodec = VideoCodecPreference.HEVC;
    public VideoCodecPreference CustomVideoCodec
    {
        get => _customVideoCodec;
        set { _customVideoCodec = value; OnPropertyChanged(); }
    }

    private TargetSizeMode _customTargetSizeMode = TargetSizeMode.Off;
    public TargetSizeMode CustomTargetSizeMode
    {
        get => _customTargetSizeMode;
        set { _customTargetSizeMode = value; OnPropertyChanged(); OnPropertyChanged(nameof(CustomSettingSummary)); }
    }

    private double _customTargetSizeMB = 25.0;
    public double CustomTargetSizeMB
    {
        get => _customTargetSizeMB;
        set { _customTargetSizeMB = value; OnPropertyChanged(); OnPropertyChanged(nameof(CustomSettingSummary)); }
    }

    private bool _stripMetadata = true;
    public bool StripMetadata
    {
        get => _stripMetadata;
        set { _stripMetadata = value; OnPropertyChanged(); }
    }

    public string FileName => Path.GetFileName(FilePath);
    public string Extension => Path.GetExtension(FilePath).TrimStart('.').ToUpperInvariant();

    public string FormattedOriginalSize
    {
        get
        {
            if (OriginalSize < 1024) return $"{OriginalSize} B";
            if (OriginalSize < 1024 * 1024) return $"{OriginalSize / 1024.0:F1} KB";
            if (OriginalSize < 1024 * 1024 * 1024) return $"{OriginalSize / (1024.0 * 1024):F1} MB";
            return $"{OriginalSize / (1024.0 * 1024 * 1024):F2} GB";
        }
    }

    public string CustomSettingSummary
    {
        get
        {
            var parts = new System.Collections.Generic.List<string>();
            if (CustomTargetSizeMode != TargetSizeMode.Off)
            {
                var targetBytes = CustomTargetSizeMode.GetTargetBytes(CustomTargetSizeMB);
                if (targetBytes.HasValue)
                {
                    parts.Add($"{targetBytes.Value / (1024 * 1024)}MB Limit");
                }
            }
            parts.Add($"{(int)(CustomQuality * 100)}% Q");
            if (CustomResolutionScale < 0.99)
            {
                parts.Add($"{(int)(CustomResolutionScale * 100)}% Scale");
            }
            return string.Join(" • ", parts);
        }
    }

    public StagedQueueItem(string filePath, long originalSize, MediaType mediaType, CompressionConfiguration baseConfig)
    {
        FilePath = filePath;
        OriginalSize = originalSize;
        MediaType = mediaType;

        CustomQuality = mediaType switch
        {
            MediaType.Image => baseConfig.ImageQuality,
            MediaType.Video => baseConfig.VideoQuality,
            _ => 0.80
        };

        CustomResolutionScale = mediaType switch
        {
            MediaType.Image => baseConfig.ImageResolutionScale,
            MediaType.Video => baseConfig.VideoResolutionScale,
            _ => 1.0
        };

        CustomImageFormat = baseConfig.ImageFormatPolicy;
        CustomVideoCodec = baseConfig.VideoCodec;
        CustomTargetSizeMode = baseConfig.TargetSizeMode;
        CustomTargetSizeMB = baseConfig.CustomTargetSizeMB;
        StripMetadata = baseConfig.StripMetadata;
    }

    public CompressionConfiguration BuildConfiguration(CompressionConfiguration baseConfig)
    {
        var config = baseConfig.Clone();
        config.ImageQuality = CustomQuality;
        config.ImageResolutionScale = CustomResolutionScale;
        config.ImageFormatPolicy = CustomImageFormat;
        config.VideoQuality = CustomQuality;
        config.VideoResolutionScale = CustomResolutionScale;
        config.VideoCodec = CustomVideoCodec;
        config.TargetSizeMode = CustomTargetSizeMode;
        config.CustomTargetSizeMB = CustomTargetSizeMB;
        config.StripMetadata = StripMetadata;
        return config;
    }

    public event PropertyChangedEventHandler? PropertyChanged;
    protected void OnPropertyChanged([CallerMemberName] string? name = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}

File: macOS/Sources/SqueezeBar/Models/StagedQueueItem.swift
import Foundation
import SwiftUI
import Combine

// MARK: - Staged Queue Item with Per-File Compression Settings
public final class StagedQueueItem: Identifiable, ObservableObject {
    public let id: UUID
    public let fileURL: URL
    public let originalSize: Int64
    public let mediaType: MediaType
    
    // Per-file individual settings
    @Published public var customQuality: Double
    @Published public var customResolutionScale: Double
    @Published public var customImageFormat: ImageFormatPolicy
    @Published public var customVideoCodec: VideoCodecPreference
    @Published public var customVideoFramerate: VideoFramerateOption
    @Published public var customVideoRemoveAudio: Bool
    @Published public var customAudioBitrate: AudioBitratePreference
    @Published public var customPDFDPI: PDFDPIOption
    @Published public var customPDFImageQuality: Double
    @Published public var customPDFGrayscale: Bool
    @Published public var customTargetSizeMode: TargetSizeMode
    @Published public var customTargetSizeMB: Double
    @Published public var stripMetadata: Bool
    
    public init(
        id: UUID = UUID(),
        fileURL: URL,
        originalSize: Int64,
        mediaType: MediaType,
        baseConfig: CompressionConfiguration
    ) {
        self.id = id
        self.fileURL = fileURL
        self.originalSize = originalSize
        self.mediaType = mediaType
        
        switch mediaType {
        case .image:
            self.customQuality = baseConfig.imageQuality
            self.customResolutionScale = baseConfig.imageResolutionScale
        case .video:
            self.customQuality = baseConfig.videoQuality
            self.customResolutionScale = baseConfig.videoResolutionScale
        case .pdf:
            self.customQuality = baseConfig.pdfImageQuality
            self.customResolutionScale = 1.0
        default:
            self.customQuality = 0.80
            self.customResolutionScale = 1.0
        }
        
        self.customImageFormat = baseConfig.imageFormatPolicy
        self.customVideoCodec = baseConfig.videoCodec
        self.customVideoFramerate = baseConfig.videoFramerate
        self.customVideoRemoveAudio = baseConfig.videoRemoveAudio
        self.customAudioBitrate = baseConfig.audioBitrate
        self.customPDFDPI = baseConfig.pdfDPI
        self.customPDFImageQuality = baseConfig.pdfImageQuality
        self.customPDFGrayscale = baseConfig.pdfGrayscale
        self.customTargetSizeMode = baseConfig.targetSizeMode
        self.customTargetSizeMB = baseConfig.customTargetSizeMB
        self.stripMetadata = baseConfig.stripMetadata
    }
    
    public var fileName: String {
        fileURL.lastPathComponent
    }
    
    public var formattedOriginalSize: String {
        ByteCountFormatter.string(fromByteCount: originalSize, countStyle: .file)
    }
    
    public var formatExtension: String {
        fileURL.pathExtension.uppercased()
    }
    
    public func buildConfiguration(baseConfig: CompressionConfiguration) -> CompressionConfiguration {
        var config = baseConfig
        
        switch mediaType {
        case .image:
            config.imageQuality = customQuality
            config.imageResolutionScale = customResolutionScale
            config.imageFormatPolicy = customImageFormat
        case .video:
            config.videoQuality = customQuality
            config.videoResolutionScale = customResolutionScale
            config.videoCodec = customVideoCodec
            config.videoFramerate = customVideoFramerate
            config.videoRemoveAudio = customVideoRemoveAudio
        case .audio:
            config.audioBitrate = customAudioBitrate
        case .pdf:
            config.pdfDPI = customPDFDPI
            config.pdfImageQuality = customPDFImageQuality
            config.pdfGrayscale = customPDFGrayscale
        case .unsupported:
            break
        }
        
        config.targetSizeMode = customTargetSizeMode
        config.customTargetSizeMB = customTargetSizeMB
        config.stripMetadata = stripMetadata
        
        return config
    }
}
